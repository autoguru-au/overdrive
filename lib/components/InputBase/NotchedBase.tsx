import { warning } from '@autoguru/utilities';
import cx from 'clsx';
import React, { FunctionComponent, useEffect, useRef, useState } from 'react';
import styles from './style.scss';

const ACTIVE_SCALING_FACTOR = 0.7777;
const ACTIVE_PADDING_ADDED = 8;
const NOTCHED_WARN_WIDTH = 170;

export interface IProps {
	id: string;
	placeholder: string;
	isEmpty: boolean;
	isActive: boolean;
}

export const NotchedBase: FunctionComponent<IProps> = ({
	id,
	placeholder,
	isEmpty,
	isActive,
	children,
}) => {
	const labelRef = useRef<HTMLLabelElement>(null);
	const [labelWidth, setLabelWidth] = useState<number>(0);

	useEffect(() => {
		if (labelRef.current) {
			setLabelWidth(labelRef.current.offsetWidth);
		}
	}, [labelRef.current, placeholder]);

	const notchedWidth =
		Math.round(
			(labelWidth * ACTIVE_SCALING_FACTOR + ACTIVE_PADDING_ADDED) * 100
		) / 100;

	// TODO: This will double warn, when placeholder gets updated, the setLabelWidth will trigger a double render
	warning(
		notchedWidth > NOTCHED_WARN_WIDTH,
		`The placeholder cannot exceed ${NOTCHED_WARN_WIDTH}px or roughly 20 characters, as it needs to fit into 280px container.`
	);

	return (
		<div
			className={cx(styles.notchedBase, {
				[styles.notchedBaseShift]: !isEmpty,
				[styles.notchedBaseActive]: isActive,
			})}>
			{children}
			<div className={styles.notchedBaseBorder}>
				<div className={styles.notchedBaseBorderLeading} />
				<div
					className={styles.notchedBaseBorderNotch}
					style={{ width: !isEmpty ? notchedWidth : 0 }}>
					<label
						ref={labelRef}
						htmlFor={id}
						className={styles.notchedBasePlaceholder}>
						{placeholder}
					</label>
				</div>
				<div className={styles.notchedBaseBorderTrailing} />
			</div>
		</div>
	);
};
