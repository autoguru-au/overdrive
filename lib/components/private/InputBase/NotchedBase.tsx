import clsx from 'clsx';
import React, { type FunctionComponent, type ReactNode, useRef } from 'react';

import { Box } from '../../Box';

import * as styles from './NotchedBase.css';
import type { BordersAttach, BordersMerged } from './NotchedBase.css';
import type { InputSize } from './withEnhancedInput.css';

export interface Props {
	id: string;
	placeholder: string;
	isEmpty: boolean;
	disabled: boolean;
	prefixed: boolean;
	borderColourClassName: string;
	placeholderColourClassName: string;
	notch?: boolean;
	size: InputSize;
	className?: string;
	children?: ReactNode;
	attach?: BordersAttach;
	borderMerged?: BordersMerged;
	isFocused?: boolean;
	isHovered?: boolean;
}

export const NotchedBase: FunctionComponent<Props> = ({
	id,
	placeholder,
	isEmpty,
	disabled,
	prefixed,
	size,
	children,
	notch = true,
	borderColourClassName,
	placeholderColourClassName,
	className = '',
	attach = 'NONE',
	borderMerged = 'NONE',
	isFocused = false,
	isHovered = false,
}) => {
	const labelRef = useRef<HTMLLabelElement>(null);

	const shouldMerge = !isFocused && !isHovered;

	return (
		<Box
			position="relative"
			width="full"
			padding="none"
			className={clsx(
				styles.root,
				!notch && [styles.borders.complete, borderColourClassName],
				className,
				{
					[styles.bordersMerged.complete.NONE]:
						!notch &&
						shouldMerge &&
						(!borderMerged || borderMerged === 'NONE'),
					[styles.bordersMerged.complete.LEFT]:
						!notch && shouldMerge && borderMerged === 'LEFT',
					[styles.bordersMerged.complete.TOP]:
						!notch && shouldMerge && borderMerged === 'TOP',
					[styles.bordersMerged.complete.RIGHT]:
						!notch && shouldMerge && borderMerged === 'RIGHT',
					[styles.bordersMerged.complete.BOTTOM]:
						!notch && shouldMerge && borderMerged === 'BOTTOM',
					[styles.bordersMerged.complete.ALL]:
						!notch && shouldMerge && borderMerged === 'ALL',
					[styles.bordersAttach.complete.NONE]:
						!notch && attach === 'NONE',
					[styles.bordersAttach.complete.LEFT]:
						!notch && attach === 'LEFT',
					[styles.bordersAttach.complete.RIGHT]:
						!notch && attach === 'RIGHT',
					[styles.bordersAttach.complete.TOP]:
						!notch && attach === 'TOP',
					[styles.bordersAttach.complete.BOTTOM]:
						!notch && attach === 'BOTTOM',
					[styles.bordersAttach.complete.ALL]:
						!notch && attach === 'ALL',
				},
			)}
		>
			{children}
			{notch && (
				<Box
					className={clsx(styles.borders.root.default, {
						[styles.borders.root.disabled]: disabled,
					})}
					pointerEvents="none"
					position="absolute"
					display="flex"
					width="full"
					height="full"
					textAlign="left"
				>
					<div
						className={clsx(
							styles.borders.leading,
							borderColourClassName,
							{
								[styles.bordersMerged.complete.LEFT]:
									shouldMerge && borderMerged === 'LEFT',
								[styles.bordersMerged.complete.TOP]:
									shouldMerge && borderMerged === 'TOP',
								[styles.bordersMerged.complete.BOTTOM]:
									shouldMerge && borderMerged === 'BOTTOM',
								[styles.bordersMerged.complete.ALL]:
									shouldMerge && borderMerged === 'ALL',
								[styles.bordersAttach.flatCorners.TOP_LEFT]:
									attach === 'LEFT' ||
									attach === 'TOP' ||
									attach === 'ALL',
								[styles.bordersAttach.flatCorners.BOTTOM_LEFT]:
									attach === 'LEFT' ||
									attach === 'BOTTOM' ||
									attach === 'ALL',
								[styles.borders.root.large]: size === 'large',
							},
						)}
					/>
					<div
						className={clsx(
							styles.borders.middle,
							borderColourClassName,
						)}
					>
						<Box
							display={
								isEmpty || placeholder.length === 0
									? 'none'
									: 'block'
							}
							paddingX="2"
							className={[
								styles.labelStyle.base,
								styles.labelStyle[size],
								styles.notchGapPlaceholder.base,
								styles.notchGapPlaceholder[size],
							]}
						>
							{placeholder}
						</Box>
						<Box
							ref={labelRef}
							as="label"
							pointerEvents="none"
							htmlFor={id}
							position="absolute"
							display="inline"
							margin="none"
							padding="none"
							className={clsx(
								styles.placeholder.default,
								placeholderColourClassName,
								styles.labelStyle.base,
								styles.labelStyle[size],
								{
									[styles.bordersMerged.complete.TOP]:
										shouldMerge && borderMerged === 'TOP',
									[styles.bordersMerged.complete.BOTTOM]:
										shouldMerge &&
										borderMerged === 'BOTTOM',
									[styles.bordersMerged.complete.ALL]:
										shouldMerge && borderMerged === 'ALL',
									[styles.placeholder.mutedLabelStyles]:
										isEmpty,
									[styles.placeholder.disabled]: disabled,
									[styles.placeholderPlacement[size].default]:
										isEmpty && !prefixed,
									[styles.placeholderPlacement[size]
										.defaultPrefixed]: isEmpty && prefixed,
									[styles.placeholderPlacement[size].shifted]:
										!isEmpty,
								},
							)}
						>
							{placeholder}
						</Box>
					</div>
					<div
						className={clsx(
							styles.borders.trailing,
							borderColourClassName,
							{
								[styles.bordersMerged.complete.RIGHT]:
									shouldMerge && borderMerged === 'RIGHT',
								[styles.bordersMerged.complete.TOP]:
									shouldMerge && borderMerged === 'TOP',
								[styles.bordersMerged.complete.BOTTOM]:
									shouldMerge && borderMerged === 'BOTTOM',
								[styles.bordersMerged.complete.ALL]:
									shouldMerge && borderMerged === 'ALL',
								[styles.bordersAttach.flatCorners.TOP_RIGHT]:
									attach === 'RIGHT' ||
									attach === 'TOP' ||
									attach === 'ALL',
								[styles.bordersAttach.flatCorners.BOTTOM_RIGHT]:
									attach === 'RIGHT' ||
									attach === 'BOTTOM' ||
									attach === 'ALL',
								[styles.borders.root.large]: size === 'large',
							},
						)}
					/>
				</Box>
			)}
		</Box>
	);
};
