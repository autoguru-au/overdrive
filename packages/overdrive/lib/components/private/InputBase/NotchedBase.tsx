import clsx from 'clsx';
import * as React from 'react';
import { FunctionComponent, useEffect, useRef, useState } from 'react';
import { useStyles } from 'react-treat';

import { Box } from '../../Box';
import { useTextStyles } from '../../Text';
import * as styleRefs from './NotchedBase.treat';

const ACTIVE_SCALING_FACTOR = 0.7777;
const ACTIVE_PADDING_ADDED = 16;
const ROUGH_WIDTH_PER_CHARACTER = 10.2;

export interface Props {
	id: string;
	placeholder: string;
	isEmpty: boolean;
	disabled: boolean;
	prefixed: boolean;
	borderColourClassName: string;
	placeholderColourClassName: string;
	notch?: boolean;
	className?: string;
}

export const NotchedBase: FunctionComponent<Props> = ({
	id,
	placeholder,
	isEmpty,
	disabled,
	prefixed,
	children,
	notch = true,
	borderColourClassName,
	placeholderColourClassName,
	className = '',
}) => {
	const styles = useStyles(styleRefs);
	const labelStyles = useTextStyles({
		noWrap: true,
		size: '4',
	});
	const labelRef = useRef<HTMLLabelElement>(null);
	const [labelWidth, setLabelWidth] = useState<number>(
		getNotchedComputedWidthForWidth(
			placeholder.length * ROUGH_WIDTH_PER_CHARACTER,
		),
	);

	useEffect(() => {
		if (labelRef.current) {
			setLabelWidth(labelRef.current.offsetWidth);
		}
	}, [placeholder]);

	const notchedWidth = getNotchedComputedWidthForWidth(labelWidth);

	return (
		<Box
			position="relative"
			width="full"
			padding="none"
			className={clsx(
				styles.root,
				!notch && [styles.borders.complete, borderColourClassName],
				className,
			)}>
			{children}
			{notch && (
				<Box
					className={clsx(styles.borders.root.default, {
						[styles.borders.root.disabled]: disabled,
					})}
					pointerEvents="none"
					position="absolute"
					display="flex"
					width="full"
					height="full"
					textAlign="left">
					<div
						className={clsx(
							styles.borders.leading,
							borderColourClassName,
						)}
					/>
					<div
						className={clsx(
							styles.borders.middle,
							borderColourClassName,
						)}
						style={{ width: isEmpty ? 0 : notchedWidth }}>
						<Box
							ref={labelRef}
							is="label"
							pointerEvents="none"
							htmlFor={id}
							position="absolute"
							display="inline"
							margin="none"
							padding="none"
							className={clsx(
								styles.placeholder.default,
								placeholderColourClassName,
								labelStyles,
								{
									[styles.placeholder.mutedLabelStyles]:
										isEmpty || disabled,
									[styles.placeholderPlacement.default]:
										isEmpty && !prefixed,
									[styles.placeholderPlacement
										.defaultPrefixed]: isEmpty && prefixed,
									[styles.placeholderPlacement
										.shifted]: !isEmpty,
								},
							)}>
							{placeholder}
						</Box>
					</div>
					<div
						className={clsx(
							styles.borders.trailing,
							borderColourClassName,
						)}
					/>
				</Box>
			)}
		</Box>
	);
};

function getNotchedComputedWidthForWidth(width: number): number {
	return (
		Math.round(
			(width * ACTIVE_SCALING_FACTOR + ACTIVE_PADDING_ADDED) * 100,
		) / 100
	);
}
