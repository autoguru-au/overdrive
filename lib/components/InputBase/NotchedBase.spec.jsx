import React from 'react';
import { render } from '@testing-library/react';
import { NotchedBase } from './NotchedBase';

describe('<NotchedBase />', () => {
	it('should not throw', () => {
		expect(() =>
			render(<NotchedBase placeholder="placeholder something" />),
		).not.toThrow();
	});

	it('should pass on className to dom element', () => {
		expect(
			render(
				<NotchedBase
					className="notched-class"
					placeholder="placeholder something"
				/>,
			).container.firstChild,
		).toHaveClass('notched-class');
	});

	it('should match snapshot for default notch', () => {
		expect(
			render(<NotchedBase placeholder="placeholder something" />)
				.container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot for dirty notch', () => {
		expect(
			render(<NotchedBase isDirty placeholder="placeholder something" />)
				.container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot for active notch', () => {
		expect(
			render(<NotchedBase isActive placeholder="placeholder something" />)
				.container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot for dirty and active notch', () => {
		expect(
			render(
				<NotchedBase
					isActive
					isDirty
					placeholder="placeholder something"
				/>,
			).container.firstChild,
		).toMatchSnapshot();
	});

	it('should display children', () => {
		expect(
			render(
				<NotchedBase placeholder="placeholder something">
					something children
				</NotchedBase>,
			).container,
		).toHaveTextContent('something children');
	});

	it('should display placeholder text inside a label element', () => {
		expect(
			render(
				<NotchedBase
					placeholder="placeholder something"
					className="notched-class"
				/>,
			).container.querySelector('label'),
		).toHaveTextContent('placeholder something');
	});

	it('should add shifted classname to dom element when isEmpty is false', () => {
		expect(
			render(
				<NotchedBase
					isEmpty={false}
					className="notched-class"
					placeholder="placeholder something"
				/>,
			).container.firstChild,
		).toHaveClass('notchedBaseShift');
	});

	it('should not shifted classname to dom element when isEmpty is true', () => {
		expect(
			render(
				<NotchedBase
					isEmpty
					className="notched-class"
					placeholder="placeholder something"
				/>,
			).container.firstChild,
		).not.toHaveClass('notchedBaseShift');
	});

	it('should add active style to dom element when isActive prop is true', () => {
		expect(
			render(
				<NotchedBase
					isActive
					className="notched-class"
					placeholder="placeholder something"
				/>,
			).container.firstChild,
		).toHaveClass('notchedBaseActive');
	});
});
