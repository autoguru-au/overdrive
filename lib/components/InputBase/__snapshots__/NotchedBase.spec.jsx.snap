// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotchedBase /> should match snapshot for active notch 1`] = `
<div
  class="base root_base"
>
  <div
    class="borders_root_default_base"
  >
    <div
      class="borderDefaults_base borders_leading_base"
    />
    <div
      class="borderDefaults_base borders_middle_base"
      style="width: 16px;"
    >
      <label
        class="placeholder_default_base placeholderPlacement_shifted_base"
      >
        placeholder something
      </label>
    </div>
    <div
      class="borderDefaults_base borders_trailing_base"
    />
  </div>
</div>
`;

exports[`<NotchedBase /> should match snapshot for default notch 1`] = `
<div
  class="base root_base"
>
  <div
    class="borders_root_default_base"
  >
    <div
      class="borderDefaults_base borders_leading_base"
    />
    <div
      class="borderDefaults_base borders_middle_base"
      style="width: 16px;"
    >
      <label
        class="placeholder_default_base placeholderPlacement_shifted_base"
      >
        placeholder something
      </label>
    </div>
    <div
      class="borderDefaults_base borders_trailing_base"
    />
  </div>
</div>
`;

exports[`<NotchedBase /> should match snapshot for dirty and active notch 1`] = `
<div
  class="base root_base"
>
  <div
    class="borders_root_default_base"
  >
    <div
      class="borderDefaults_base borders_leading_base"
    />
    <div
      class="borderDefaults_base borders_middle_base"
      style="width: 16px;"
    >
      <label
        class="placeholder_default_base placeholderPlacement_shifted_base"
      >
        placeholder something
      </label>
    </div>
    <div
      class="borderDefaults_base borders_trailing_base"
    />
  </div>
</div>
`;

exports[`<NotchedBase /> should match snapshot for dirty notch 1`] = `
<div
  class="base root_base"
>
  <div
    class="borders_root_default_base"
  >
    <div
      class="borderDefaults_base borders_leading_base"
    />
    <div
      class="borderDefaults_base borders_middle_base"
      style="width: 16px;"
    >
      <label
        class="placeholder_default_base placeholderPlacement_shifted_base"
      >
        placeholder something
      </label>
    </div>
    <div
      class="borderDefaults_base borders_trailing_base"
    />
  </div>
</div>
`;
